src/custom/components/Tokens/TokensTableRow.tsx

```diff
@@ -24,13 +35,106 @@ type DataRowParams = {
   balance?: CurrencyAmount<Token> | undefined
   handleBuy: (token: Token) => void
   handleSell: (token: Token) => void
+  closeModal: () => void
+  openTransactionConfirmationModal: (message: string, operationType: OperationType) => void
+  toggleWalletModal: () => void
 }
 
-const DataRow = ({ tokenData, index, balance, handleBuy, handleSell }: DataRowParams) => {
+const DataRow = ({
+  tokenData,
+  index,
+  balance,
+  handleBuy,
+  handleSell,
+  closeModal,
+  openTransactionConfirmationModal,
+  toggleWalletModal,
+}: DataRowParams) => {
+  const { account, chainId } = useActiveWeb3React()
+
   const theme = useTheme()
   const hasBalance = balance?.greaterThan(0)
   const formattedBalance = formatSmart(balance) || 0
 
+  // allowance
+  const spender = chainId ? GP_VAULT_RELAYER[chainId] : undefined
+  const currentAllowance = useTokenAllowance(tokenData, account ?? undefined, spender)
+
+  // approve
+  const [approving, setApproving] = useState(false)
+
+  const { handleSetError, handleCloseError } = useErrorModal()
+
+  const { approvalState, approve } = useApproveCallbackFromBalance({
+    openTransactionConfirmationModal,
+    closeModals: closeModal,
+    token: tokenData,
+    balance,
+  })
+
+  const prevApprovalState = usePrevious(approvalState)
+
+  const handleApprove = useCallback(async () => {
+    handleCloseError()
+
+    if (!account) {
+      toggleWalletModal()
+      return
+    }
+
+    try {
+      setApproving(true)
+      const summary = `Approve ${tokenData?.symbol || 'token'}`
+      await approve({ modalMessage: summary, transactionSummary: summary })
```

Comment:
we'll also need anayltics here. I feel like this is not DRY code and we could make a hook out of this with the analytics involved (GA) - not for this PR

Reply:
Added a TODO comment :+1: